proposals/bip-002.md

```diff
@@ -0,0 +1,93 @@
+# Explicit Annotated status and completeness information
+
+| Field | Value |
+| --- | --- |
+| BIP | 2 |
+| Title | Explicit Annotated status and completeness information |
+| Author | Juanjo Álvarez |
+| Status | Draft |
+| Created | 2017-06-01 |
+| Updated | 2017-06-01 |
+
+## Abstract
+
+This BIP proposed a new official (as in defined in the Node struct and
+documented) "AnnotationStatus" property that would mark if the role has been
+annotated in any way and the quality of the current annotation.
+
+## Rationale
+
+At the current stage of the Babelfish project, we strive for genericity in the list
+of UAST roles and not adding very language-specific features into it until we can
+have a wider view about how to integrate or compose these roles for covering as
+much ground as possible. 
+
+This is a worthy objective but has the problem of leaving the nodes related with
+these features unannotated or with annotations that don't fully express all
+the semantic information that the language source code or the native AST have.
+
+This BIP proposes to add a new property in the Nodes called "AnnotationStatus"
+with the possible values of "Annotated", "Unannotated" and "Incomplete". This will
+make the UAST user have a wider view of the completeness of the information
+provided by it on specific nodes and also could help debug drivers and provide a
+completeness summary, with the added benefit of providing a numerically
+constatable measure of the completeness of every developed driver.
+
+
+The Unannotated value for that property could be arguably less needed because it
+can be inferred by the lack of any role in a specific node, but is proposed for
+completeness. The "Incomplete" value, on the other hand, used in an additive way
+with other roles in the UAST, would provide worthy information about the lack of
+semantic information of the roles provided, which could be used by an UAST user to
+treat those nodes in a different way or, when knowing the specific language, look
+at the native AST information that the UAST nodes carry as properties.
+
+Example:
+
+If a version of the UAST specification didn't have a role that could be matched
+for a "coroutine" type function supported by some language, the UAST could still
+mark those nodes as "FunctionDef", but without additional information, this would be
+providing incomplete and even misleading information to UAST users that could
+treat coroutines differently (for example when making stack call deep analysis).
+
+## Specification
+
+This would modify the current specification by adding a new `AnnotationStatus`
+property into the `Node` struct (and the UAST nodes) and a `const` enum with
+the possible values.
+
+## Alternatives
+
+Instead of a new property, "Unannotated" and "Incomplete" could be UAST "meta"
```

Comment:
That was my first design, but IMHO it adds roles of a new category that doesn't provide semantic information and since we're already using properties for some other meta information (like position or original AST roles) I think properties are the natural fit for this status (that is also metainformation).

Also, we're trying to not grow much the # of UAST roles for the initial version and this would add two more.